Add support to log customized tags to runs created by autolog

thinkall opened the pull request:
<!-- 🚨 We recommend pull requests be filed from a non-master branch on a repository fork (e.g. <username>:fix-xxx). 🚨 -->

## Related Issues/PRs

Resolve #9064

## What changes are proposed in this pull request?

Added a parameter `extra_tags` to all `autolog` functions, and passed it to `safe_patch` function in `mlflow.utils.autologging_utils.safety.py`.

With this `extra_tags`, users can log customized tags to runs created by autolog.

## How is this patch tested?

<!--
If you're unsure about what to test, where to add tests, or how to run tests, please feel free to ask.
-->

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

<!--
Please describe how you confirmed the proposed feature/bug-fix/change works here. For example, if you fixed an MLflow client API, you could attach the code that didn't work prior to the fix but works now, or if you added a new feature on MLflow UI, you could attach a video that demonstrates the feature.
-->

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Users can set the parameter `extra_tags` in `mlflow.autolog(extra_tags={"test_tag": "autolog"})` and all the other specific autolog  functions such as `mlflow.sklearn.autolog(extra_tags={"test_tag": "sklearn_autolog"})` to log the `extra_tags` into runs created by autolog.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

thinkall committed:
Add support to log customized tags to runs created by autolog #9064

Signed-off-by: Li Jiang <bnujli@gmail.com>

thinkall committed:
Update comments

Signed-off-by: Li Jiang <bnujli@gmail.com>

mlflow-automation commented:
<!-- documentation preview -->


Documentation preview for 9e99f8adfd2b810df8f275a783faa53282a1acc9 will be available [here](https://output.circle-artifacts.com/output/job/5839d3f7-97e2-4d07-8ddd-93fc610d55e6/artifacts/0/docs/build/html/index.html) when [this CircleCI job](https://circleci.com/gh/mlflow/mlflow/57404) completes successfully.

<details>
<summary>More info</summary>

- Ignore this comment if this PR does not change the documentation.
- It takes a few minutes for the preview to be available.
- The preview is updated when a new commit is pushed to this PR.
- This comment was created by https://github.com/mlflow/mlflow/actions/runs/5594443530.

</details>

thinkall committed:
Remove test_autolog_extra_tags.py, move tests to corresponding files to fix test failures

Signed-off-by: Li Jiang <bnujli@gmail.com>

thinkall committed:
Fix pytorch test for extra_tags

Signed-off-by: Li Jiang <bnujli@gmail.com>

thinkall committed:
Fix transformers test for extra_tags

Signed-off-by: Li Jiang <bnujli@gmail.com>

thinkall committed:
Remove transformers test for extra_tags as manage_run are all False

Signed-off-by: Li Jiang <bnujli@gmail.com>

thinkall committed:
Fix sklearn test for extra_tags

Signed-off-by: Li Jiang <bnujli@gmail.com>

thinkall commented:
Hi @serena-ruan , could you help review the PR? Thanks.

serena-ruan reviewed (commented):
@thinkall Overall looks great! Left a few comments :)

I think it could be more clear that we split tags test.

thinkall committed:
Merge branch 'master' into add_custom_tags_autolog

thinkall committed:
Add warning msg and built-in tag checking

Signed-off-by: Li Jiang <bnujli@gmail.com>

thinkall committed:
Move all tests for extra_tags to separate tests

Signed-off-by: Li Jiang <bnujli@gmail.com>

thinkall commented:
> @thinkall Overall looks great! Left a few comments :)
> 
> I think it could be more clear that we split tags test.

Thank you Serana, I've split all the tests for extra_tags. Warning messages and built-in tags checking have been included as well.

thinkall committed:
Fix format

Signed-off-by: Li Jiang <bnujli@gmail.com>

serena-ruan commented on mlflow/utils/autologging_utils/safety.py:
```diff
@@ -304,14 +309,23 @@ class definition or a function object. If it is a function object, the
+                _logger.warning("Tag %s is being ignored because it conflicts with a built-in tag.", MLFLOW_AUTOLOGGING)
+            tags.update(extra_tags)
+        elif extra_tags is not None:
+            _logger.warning("The parameter `extra_tags` is being ignored because it is expected to be a dictionary, but the provided value is %s", type(extra_tags))
```
```suggestion
        tags = {MLFLOW_AUTOLOGGING: autologging_integration}
        if extra_tags:
            if isinstance(extra_tags, dict):
                if MLFLOW_AUTOLOGGING in extra_tags:
                    extra_tags.pop(MLFLOW_AUTOLOGGING)
                    _logger.warning(f"Tag `{MLFLOW_AUTOLOGGING}` is ignored as it is a reserved tag by MLflow autologging.")
                tags.update(extra_tags)
            else:
                raise MlflowException.invalid_parameter_value(f"Invalid `extra_tags` type: expecting dictionary, received `{type(extra_tags).__name__}`")
```

Sorry I think it makes more sense to throw an exception if it's not a dictionary :)

serena-ruan commented on tests/fastai/test_fastai_autolog.py:
```diff
@@ -84,6 +84,15 @@ def test_fastai_autolog_ends_auto_created_run(iris_data, fit_variant):
+    model.fit_one_cycle(1)
+    
+    run = mlflow.last_active_run()
+    assert run.data.tags["test_tag"] == "fastai_autolog"
```
```suggestion
    assert run.data.tags["test_tag"] == "fastai_autolog"
    assert run.data.tags[MLFLOW_AUTOLOGGING] == "fastai"
```

serena-ruan replied:
Can we add MLFLOW_AUTOLOGGING tag check for all flavors as well? Since you've already add tags check :)

thinkall replied:
Added :)

thinkall committed:
Merge branch 'mlflow:master' into add_custom_tags_autolog

thinkall committed:
Add raise error and more tests for MLFLOW_AUTOLOGGING overridden

Signed-off-by: Li Jiang <bnujli@gmail.com>

thinkall committed:
Merge master

Signed-off-by: Li Jiang <bnujli@gmail.com>

serena-ruan commented:
LGTM! @harupy You wanna take a look? 😃

thinkall commented:
> 

Thank you @serena-ruan, and hi, @harupy ! Just a gentle reminder to take a look :-)

serena-ruan reviewed (approved):
LGTM
